packages/@aws-cdk/aws-docdb/lib/endpoint.ts

```diff
@@ -0,0 +1,81 @@
+import { Token } from '@aws-cdk/core';
+
+/**
+ * Connection endpoint of a database cluster or instance
+ *
+ * Consists of a combination of hostname and port.
+ */
+export class Endpoint {
+  /**
+   * The minimum port value
+   */
+  private static readonly MIN_PORT = 1;
+
+  /**
+   * The maximum port value
+   */
+  private static readonly MAX_PORT = 65535;
+
+  /**
+   * Determines if a port is valid
+   *
+   * @param port: The port number
+   * @returns boolean whether the port is valid
+   */
+  private static isValidPort(port: number): boolean {
+    return Number.isInteger(port) && port >= Endpoint.MIN_PORT && port <= Endpoint.MAX_PORT;
+  }
+
+  /**
+   * The hostname of the endpoint
+   */
+  public readonly hostname: string;
+
+  /**
+   * The port number of the endpoint.
+   *
+   * This can potentially be a CDK token. If you need to embed the port in a string (e.g. instance user data script),
+   * use {@link Endpoint.portAsString}.
+   */
+  public readonly port: number;
+
+  /**
+   * The combination of "HOSTNAME:PORT" for this endpoint
+   */
+  public readonly socketAddress: string;
+
+  /**
+   * Constructs an Endpoint instance.
+   *
+   * @param address - The hostname or address of the endpoint
+   * @param port - The port number of the endpoint
+   */
+  constructor(address: string, port: number) {
+    if (!Token.isUnresolved(port) && !Endpoint.isValidPort(port)) {
+      throw new Error(`Port must be an integer between [${Endpoint.MIN_PORT}, ${Endpoint.MAX_PORT}] but got: ${port}`);
+    }
+
+    this.hostname = address;
+    this.port = port;
+
+    const portDesc = Token.isUnresolved(port) ? Token.asString(port) : port;
+    this.socketAddress = `${address}:${portDesc}`;
+  }
+
+  /**
+   * Returns the port number as a string representation that can be used for embedding within other strings.
+   *
+   * This is intended to deal with CDK's token system. Numeric CDK tokens are not expanded when their string
+   * representation is embedded in a string. This function returns the port either as an unresolved string token or
+   * as a resolved string representation of the port value.
+   *
+   * @returns {string} An (un)resolved string representation of the endpoint's port number
+   */
+  public portAsString(): string {
```

Comment:
Im not a hundred percent convinced here. Can you share an example usage of string embedding that doesn't work without this?

Reply:
Consider using the port as a string in an instance's user data script.

```ts
import * as cdk from '@aws-cdk/core';
import * as ec2 from '@aws-cdk/aws-ec2';
import * as docdb from '@aws-cdk/aws-docdb';

const app = new cdk.App();
const stack = new cdk.Stack(app, 'MyStack');
const vpc = new ec2.Vpc(stack, 'Vpc');

const cluster = new docdb.DatabaseCluster(stack, 'Database', {
  masterUser: {
    username: 'docdb'
  },
  instanceProps: {
    instanceType: ec2.InstanceType.of(ec2.InstanceClass.R5, ec2.InstanceSize.LARGE),
    vpc
  },
  removalPolicy: cdk.RemovalPolicy.DESTROY
});

// Create instance to run my application
const instance = new ec2.Instance(stack, 'Instance', {
  vpc,
  instanceType: ec2.InstanceType.of(ec2.InstanceClass.T3, ec2.InstanceSize.MICRO),
  machineImage: new ec2.AmazonLinuxImage()
});

// Allow network traffic from the instance
cluster.connections.allowDefaultPortFrom(instance);

// Start my app on launch and connect to the DocumentDB cluster
instance.userData.addCommands(
  'pip install mysuperawesomeapp',
  `mysuperawesomeapp --dbhost ${cluster.clusterEndpoint.hostname} --dbport ${cluster.clusterEndpoint.portAsString()}`
);
```

CDK does not resolve numeric tokens within strings. If you use `cluster.clsuterEndpoint.port` in the userdata command string it gets left as an unresolved token.